### PR TITLE
Deduplicate module names in FCS

### PR DESCRIPTION
**Pull request by forki**
this ports the work of @alfonsogarciacaro from FCS back to VF#

/see also https://github.com/fsharp/FSharp.Compiler.Service/pull/749

### CODE REVIEW

**Comment by forki** [comment]
Jus for sake of complteness. It seems fsi.exe is fine:

image

File: tests/service/data/samename/tempet.fsproj [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<Project Sdk="FSharp.NET.Sdk;Microsoft.NET.Sdk">
```
Comment (KevinRansom) [inline tests/service/data/samename/tempet.fsproj]: Does this test anything valuable?  After all it is not using the locally built CompilerService.

**Comment by forki** [comment]
Yeah I guess the test is not correct. But the fix is. Will investigate on
Monday

Am 02.04.2017 8:09 nachm. schrieb "Kevin Ransom (msft)" <
notifications@github.com>:

*@KevinRansom* commented on this pull request.
------------------------------

In tests/service/data/samename/tempet.fsproj
<https://github.com/Microsoft/visualfsharp/pull/2772#discussion_r109315635>:

> @@ -0,0 +1,13 @@
+<Project Sdk="FSharp.NET.Sdk;Microsoft.NET.Sdk">

Does this test anything valuable? After all it is not using the locally
built CompilerService.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/Microsoft/visualfsharp/pull/2772#pullrequestreview-30410284>,
or mute the thread

**Comment by alfonsogarciacaro** [comment]
@KevinRansom The test is necessary for FCS as unfortunately the fix from #2728 didn't apply when calling `FSharpChecker.ParseAndCheckProject`. I don't know how the FCS tests are ported to visualfsharp, I can see that the tests files are in `tests/service` but the tests are not run and most of the project samples in `tests/service/data` are missing.

Please note the logic in this PR it's the same as in #2728 applied to a different place. Maybe we could leave a comment as a reminder to refactor both fixes to a common place in the future?

**Comment by forki** [comment]
@alfonsogarciacaro @KevinRansom I think the fsproj file is not used at all. the test itself is in NUnit

**Comment by forki** [comment]
ok further tracing showed that the fix actually deduped the module but it still fails in tast.fs - CombineModuleOrNamespaceTypes

so I need to do more investigation ....

**Comment by forki** [comment]
parse D:\j\workspace\release_ci_pa---3f142ccc\tests\service/data/samename/folder1/a.fs
new D:\j\workspace\release_ci_pa---3f142ccc\tests\service/data/samename/folder1/a.fs -> A
parse D:\j\workspace\release_ci_pa---3f142ccc\tests\service/data/samename/folder2/a.fs
seen D:\j\workspace\release_ci_pa---3f142ccc\tests\service/data/samename/folder2/a.fs -> A___2
add files with same name from different folders

ERROR: Two modules named 'tempet.SayA' occur in two parts of this assembly

So it's complaing about the module SayA

lol - I didn't copy the test correctly.

image

File: tests/service/data/samename/tempet.fsproj (hunk 1 above)
Comment (dsyme) [inline tests/service/data/samename/tempet.fsproj]: I'm not sure about ``tempet.fsproj``, but the tests in ``tests/service/ProjectAnalysisTests.fs`` are, I think, also being run against FSharp.LanguageService.Compiler.dll (the locally built compiler service DLL)

Comment (forki) [inline tests/service/data/samename/tempet.fsproj]: yes it's a NUnit test that loads the proj. I think it doesn't use the sdk at all

**Comment by forki** [comment]
Ok fixed. Ready to go in

File: src/fsharp/vs/IncrementalBuild.fs [hunk 2]
```diff
@@ -1328,6 +1328,14 @@ type IncrementalBuilder(tcGlobals,frameworkTcImports, nonFrameworkAssemblyInputs
     let StampFileNameTask (cache: TimeStampCache) _ctok (_m:range, filename:string, _isLastCompiland) =
         assertNotDisposed()
         cache.GetFileTimeStamp filename
+
+    // Deduplicate module names
+    let seen = Dictionary<string,Set<string>>()
```
Comment (dsyme) [inline src/fsharp/vs/IncrementalBuild.fs]: @forki This duplicates the code added to fsc.fs?  Could you extract this to CompileOptions.fs or CompileOps.fs and share it please? https://github.com/Microsoft/visualfsharp/blob/f6e160a2e4a375754bcf62fc3641b1861093ca3b/src/fsharp/fsc.fs#L1768

Comment (forki) [inline src/fsharp/vs/IncrementalBuild.fs]: done in 2fc7247

**Comment by forki** [comment]
I also send a Pull request to @alfonsogarciacaro's PR so that both are in sync

**Comment by forki** [comment]
can you please merge and release the corresponding FCS PR? It's needed for ionide ;-)

**Comment by dsyme** [comment]
yes will do

**Comment by dsyme** [comment]
done, 12.0.3

**Comment by ThisFunctionalTom** [comment]
I came across same/similar issue. [Here is the smallest repro](here: https://github.com/ThisFunctionalTom/FsxLoadBugRepro) I could come up with.

Maybe it can help.

**Pull request merged**